frontends/concrete-python/concrete/fhe/extensions/multivariate.py

```diff
@@ -0,0 +1,104 @@
+"""
+Declaration of `multivariate` extension.
+"""
+
+from copy import deepcopy
+from typing import Any, Callable, Optional, Type, Union
+
+import numpy as np
+
+from ..dtypes import BaseDataType, Float
+from ..representation import Node
+from ..tracing import ScalarAnnotation, Tracer
+from ..values import ValueDescription
+
+
+def multivariate(
+    function: Callable,
+    outputs: Optional[Union[BaseDataType, Type[ScalarAnnotation]]] = None,
+) -> Callable:
+    """
+    Wrap a multivariate function so that it is traced into a single generic node.
+
+    Args:
+        function (Callable[[Any, ...], Any]):
+            multivariate function to wrap
+
+        outputs (Optional[Union[BaseDataType, Type[ScalarAnnotation]]], default = None):
+            data type of the result, unused during compilation, required for direct definition
+
+    Returns:
+        Callable[[Union[Tracer, Any], ...], Union[Tracer, Any]]:
+            another multivariate function that can be called with Tracers as well
+    """
+
+    def wrapper(*args: Union[Tracer, Any]) -> Union[Tracer, Any]:
+        """
+        Evaluate or trace wrapped multivariate function.
+
+        Args:
+            args (Tuple[Union[Tracer, Any]]):
+                inputs of the function
+
+        Returns:
+            Union[Tracer, Any]:
+                result of tracing or evaluation
+        """
+
+        if any(isinstance(arg, Tracer) for arg in args):
+            # pylint: disable=protected-access
+            is_direct = Tracer._is_direct
+            # pylint: enable=protected-access
+
+            if is_direct:
+                if outputs is None:
+                    message = (
+                        "Multivariate extension requires "
+                        "`outputs` argument for direct circuit definition "
+                        "(e.g., fhe.multivariate(function, outputs=fhe.uint4)(x, y, z))"
+                    )
+                    raise ValueError(message)
+
+            if any(not isinstance(arg, Tracer) or not arg.output.is_encrypted for arg in args):
+                message = "Multivariate extension requires all of its inputs to be encrypted"
+                raise ValueError(message)
+
+            if any(arg.computation.properties.get("name") == "round_bit_pattern" for arg in args):
```

Comment:
What are the current road block ?

Reply:
For clear inputs:
- Clear multiplication, clear addition, maybe TLU on them, and some way to track signedness of the clear value (it's signless atm)

For rounding:
- It needs some special care, and this PR was already getting long. It can come with a subsequent PR. Though there are very weird cases, it's not that straightforward.